code/dirs.py

```diff
@@ -260,3 +307,230 @@ def makedirs_if_not_exist(directory):
+    if func(0, buf, get_csidl(app_dir), True):
+        return buf.value
+    return None
+
+def shell32_SHGetSpecialFolderPathA(func, app_dir):
+    buf = ctypes.create_string_buffer(ctypes.wintypes.MAX_PATH + 1)
+    if (func(0, buf, get_csidl(app_dir), True)):
+        return buf.value
+    return None
+
+def get_win_folders():
+
+    # First try to import ctypes to call c win32 function.
+    # win32 function is the correct way to get win folder.
+    # but python is not shipped on windows with pywin32
+    # Let's do the hard way.
+    try:
+        import ctypes
+        import ctypes.wintypes
+    except:
+        if g.debug: sys.stderr.write("WARNING: Impossible to load ctypes modules\n")
+
+    if 'ctypes' in sys.modules:
+        for func_name in win32_function:
+            try:
+                func = ctypes.windll.shell32.getattr(func_name)
+                
+                handler = globals().get("shell32_%s" % func_name)
+                
+                local_appdir = handler(func, 'local')
+                roaming_appdir = handler(func, 'roaming')
+                
+                if (roaming_appdir is not None and local_appdir is not None)
+                    return (roaming_appdir, local_appdir)
+                else:
+                    if g.debug: sys.stderr.write("WARNING: '%s' shell32 function fails for local and roaming\n" % func_name)
+            except:
+                if g.debug: sys.stderr.write("WARNING: Impossible to use '%s' shell32 function\n" % func_name)
+
+    # ctypes doesn't work, let find the APPDATA environnement variables.
+    if os.environ.has_key("APPDATA") and os.environ.has_key("LOCALAPPDATA"):
+        
+        local_appdir = os.environ.has_key("LOCALAPPDATA")
+        roaming_appdir = os.environ.has_key("APPDATA")
+
+        return (roaming_appdir, local_appdir)
+    else:
+        if g.debug: sys.stderr.write("WARNING: APPDATA and LOCALAPPDATA environment not present \n")
+
+    # Ok, we could try to get home user directories.
+    # Get it to all different way.
+    home_dirs = []
+
+    if 'ctypes' in sys.modules:
+        for func_name in win32_function:
+            try:
+                func = ctypes.windll.shell32.getattr(func_name)
+                
+                handler = globals().get("shell32_%s" % func_name)
+                
+                profiledir = handler(func, 'home')
+                
+                if profiledir is not None:
+                    home_dirs.append(profiledir)
+                else:
+                    if g.debug: sys.stderr.write("WARNING: '%s' shell32 function fails for home\n" % func_name)
+            except:
+                if g.debug: sys.stderr.write("WARNING: Impossible to use '%s' shell32 function for home\n" % func_name)
+
+    if os.environ.has_key("USERPROFILE")
+        home_dirs.append(os.environ.get("USERPROFILE"))
+    else:
+        if g.debug: sys.stderr.write("WARNING: USERPROFILE environment not present \n")
+
```

Comment:
This if-else block should either use `Y in X` (instead of `X.has_key(Y)`) or alternatively be written as:

```python3
try:
    home_dirs.append(os.environ[Y])
except KeyError:
    if g.debug:
       sys.stderr.write("WARNING: Y environment not present \n")
```